File: advanced-programming-methods/labs/toy-interpreter-java/src/model/expression/HeapReadExpression.java
package model.expression;

import exceptions.DictionaryException;
import exceptions.ExpressionException;
import exceptions.HeapException;
import model.IDictionary;
import model.IHeap;

public class HeapReadExpression extends Expression
{
    private String varName;

    /**
     * Instantiates a new Var expression.
     *
     * @param varName the var name
     */
    public HeapReadExpression(String varName)
    {
        this.varName = varName;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public int eval(IDictionary table, IHeap heap) throws ExpressionException
    {
        int address;
        try
        {
            address = table.get(varName);
        }
        catch (DictionaryException e)
        {
            throw new ExpressionException(
                String.format("Variable = '%s' is not defined", varName)
            );
        }

        try
        {
            return heap.read(address);
        }
        catch (HeapException e)
        {
            throw new ExpressionException(
                String.format("Address = %d is out of bounds", address)
            );
        }
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public String toString()
    {
        return String.format("r(%s)", varName);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public Expression cloneDeep()
    {
        return new HeapReadExpression(varName);
    }
}

File: advanced-programming-methods/labs/toy-interpreter-sharp/model/expression/HeapReadExpression.cs
using System;
using Exceptions;

namespace Model
{
    [Serializable()]
    public class HeapReadExpression : Expression
    {
        private string varName;

        public HeapReadExpression(string varName)
        {
            this.varName = varName;
        }

        /// <inheritdoc />
        public override int Eval(IDictionary table, IHeap heap)
        {
            int address;
            try
            {
                address = table.Get(varName);
            }
            catch (DictionaryException)
            {
                throw new ExpressionException(string.Format("Variable = '{0}' is not defined", varName));
            }

            try
            {
                return heap.Read(address);
            }
            catch (HeapException)
            {
                throw new ExpressionException(string.Format("Address = {0} is out of bounds", address));
            }
        }

        /// <summary> {@inheritDoc} </summary>
        public override string ToString()
        {
            return string.Format("r({0})", varName);
        }

        /// <summary> {@inheritDoc} </summary>
        public override Expression CloneDeep()
        {
            return new HeapReadExpression(varName);
        }
    }
}

